explorer/ast/ast_rtti.h

```diff
@@ -37,82 +37,51 @@ class AstRttiNodeKind : public CARBON_ENUM_BASE(AstRttiNodeKind) {
+    /* This type can be explicitly converted from the generic node kind. */   \
+    explicit C##KindTemplate(AstRttiNodeKind base_kind)                       \
+        : Base(Base::FromInt(base_kind.AsInt())) {}                           \
+                                                                              \
+    CARBON_##C##_KINDS(CARBON_IGNORE, CARBON_INLINE_ENUM_CONSTANT_DEFINITION) \
+  };                                                                          \
+  class C##Kind : public C##KindTemplate<C##Kind> {                           \
+    using C##KindTemplate<C##Kind>::C##KindTemplate;                          \
   };
-#define ENUMERATOR(E) E = AstRttiNodeKind::E.AsInt(),
-CARBON_AST_RTTI_KINDS(DEFINE_ENUM, IGNORE)
+#define DEFINE_ENUMERATOR(E) E = AstRttiNodeKind::E.AsInt(),
+CARBON_AST_RTTI_KINDS(DEFINE_KIND_ENUM, CARBON_IGNORE)
 #undef DEFINE_ENUM
-#undef ENUMERATOR
-
-// Helper to compute the number of derived classes for a class by repeatedly
-// adding one.
-#define ADD_ONE(C) +1
-#define NUMBER_OF_DERIVED_CLASSES(C) CARBON_##C##_KINDS(IGNORE, ADD_ONE)
-
-// Helper to find the first derived class of a class.
-#define EXPAND(A) A
-#define LPAREN (
-#define RPAREN )
-#define FIRST(A, ...) A
-#define COMMA_AFTER(A) A,
-#define FIRST_DERIVED_CLASS_IMPL(C) \
-  FIRST LPAREN CARBON_##C##_KINDS(IGNORE, COMMA_AFTER) RPAREN
-#define FIRST_DERIVED_CLASS(C) EXPAND(FIRST_DERIVED_CLASS_IMPL(C))
-
-// Helper to find the last derived class of a class. This works by converting
-// the list of derived classes A B C into
-//   APPLY(, DISCARD) APPLY(A, DISCARD) APPLY(B, DISCARD) APPLY(C, RETURN)
-// which then expands to
-//   DISCARD() DISCARD(A) DISCARD(B) RETURN(C)
-// then to
-//   C
-#define DISCARD(...)
-#define RETURN(A) A
-#define APPLY(E, M) M(E)
-#define LAST_HELPER(E) , DISCARD RPAREN APPLY LPAREN E
-#define LAST_DERIVED_CLASS_IMPL(C) \
-  APPLY LPAREN CARBON_##C##_KINDS(IGNORE, LAST_HELPER), RETURN RPAREN
-#define LAST_DERIVED_CLASS(C) EXPAND(LAST_DERIVED_CLASS_IMPL(C))
+#undef DEFINE_ENUMERATOR
 
 // Define InheritsFrom functions for all kinds.
-#define DEFINE_INHERITS_FROM_FUNCTION_ABSTRACT(C)                            \
-  inline bool InheritsFrom##C(Carbon::AstRttiNodeKind kind) {                \
-    constexpr auto kFirst = Carbon::AstRttiNodeKind::FIRST_DERIVED_CLASS(C); \
-    constexpr auto kLast = Carbon::AstRttiNodeKind::LAST_DERIVED_CLASS(C);   \
-    static_assert(NUMBER_OF_DERIVED_CLASSES(C) ==                            \
-                  kLast.AsInt() - kFirst.AsInt() + 1);                       \
-    return kind >= kFirst && kind <= kLast;                                  \
+#define DEFINE_INHERITS_FROM_FUNCTION_ABSTRACT(C)                             \
+  inline bool InheritsFrom##C(Carbon::AstRttiNodeKind kind) {                 \
+    return CARBON_##C##_KINDS(CARBON_IGNORE, INHERITS_FROM_CLASS_TEST) false; \
   }
+#define INHERITS_FROM_CLASS_TEST(C) kind == Carbon::AstRttiNodeKind::C##Kind ||
```

Comment:
Did you mean to undef this?

Reply:
Sorry, pushed an incomplete commit. PTAL :)